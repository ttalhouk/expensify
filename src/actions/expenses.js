import uuid from 'uuid';
import database from '../firebase/firebase';


export const addExpense = (expense) => {
  return {
    type: 'ADD_EXPENSE',
    expense
  };
}

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {
      description,
      note,
      amount,
      createdAt
    }
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const removeExpense = ({id}) => {
  return {
    type:'REMOVE_EXPENSE',
    id
  }
}

export const startRemoveExpense = ({id}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const editExpense = (id, updates) => {
  return (
    {
      type: 'EDIT_EXPENSE',
      id,
      updates
    }
  );
};
export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const setExpenses = (expenses) => {
  return {
    type: 'SET_EXPENSES',
    expenses
  }
}

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((expense) => {
            expenses.push({
              id: expense.key,
              ...expense.val()
            })
          });
        dispatch(setExpenses(expenses));
      }).catch((error) => {
        console.log(error);
      });
  }
}
