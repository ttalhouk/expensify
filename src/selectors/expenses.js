import moment from 'moment';

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const startDateMatch = startDate ? startDate.isSameOrBefore(moment(expense.createdAt), 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt), 'day') : true;
    const textMatch =  expense.description.toLowerCase().includes(text.toLowerCase());

    return textMatch && startDateMatch && endDateMatch;
  }).sort((exp1, exp2) => {
    if (sortBy === 'date') {
      return exp2.createdAt - exp1.createdAt;
    } else if (sortBy === 'amount'){
      return exp2.amount - exp1.amount;
    }
  })
}

export default getVisibleExpenses;
