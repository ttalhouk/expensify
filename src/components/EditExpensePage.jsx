import React from 'react';
import {connect} from 'react-redux';

import {startEditExpense, startRemoveExpense} from '../actions/expenses';

import ExpenseForm from './ExpenseForm.jsx';


export class EditExpensePage extends React.Component {

  onSubmitUpdates = (updates) => {
    this.props.startEditExpense(this.props.expense.id, updates);
    this.props.history.push('/');
  }
  onRemoveExpense = () => {
    this.props.startRemoveExpense({id: this.props.expense.id});
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense <span>{this.props.expense.description}</span></h1>
          </div>
        </div>
          <div className="content-container">
            <ExpenseForm
              expense={this.props.expense}
              onSubmit={this.onSubmitUpdates}
              />
            <button
              className="button button--remove"
              onClick={this.onRemoveExpense}
              >
              Remove Expense
            </button>
          </div>

      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id
    })
  }
}
const mapDispatchToProps = (dispatch) => {
  return {
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
    startEditExpense: (id, updates) => dispatch(startEditExpense(id, updates))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
