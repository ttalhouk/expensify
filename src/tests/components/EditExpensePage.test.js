import React from 'react';
import {shallow} from 'enzyme';

import {EditExpensePage} from '../../components/EditExpensePage.jsx';

import {expenses} from '../fixtures/expenses'

describe("<EditExpensePage />", () => {
  let history, onSubmitSpy, onRemoveSpy, wrapper, expense;
  beforeEach(() => {
    history = {push: jest.fn()};
    expense = expenses[1]
    onSubmitSpy = jest.fn();
    onRemoveSpy = jest.fn()
    wrapper = shallow(<EditExpensePage expense={expense} history={history} startEditExpense={onSubmitSpy} startRemoveExpense={onRemoveSpy}/>);
  })
  it('should render the page correctly', () => {
    expect(wrapper).toMatchSnapshot()
  });
  it('should call startEditExpense when submitting form', () => {

    wrapper.find('ExpenseForm').prop('onSubmit')(expense);
    expect(onSubmitSpy).toHaveBeenLastCalledWith(expense.id, expense);
    expect(history.push).toHaveBeenLastCalledWith('/');
  });
  it('should call startRemoveExpense when button clicked',() => {
    wrapper.find('button').simulate('click');
    expect(onRemoveSpy).toHaveBeenLastCalledWith({id: expense.id});
    expect(history.push).toHaveBeenLastCalledWith('/');
  })
});
