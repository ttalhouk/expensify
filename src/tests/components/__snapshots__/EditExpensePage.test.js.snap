// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditExpensePage /> should render the page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense 
        <span>
          water bill
        </span>
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 40,
          "createdAt": 0,
          "description": "water bill",
          "id": "2",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--remove"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
